File: backend/app/api/v1/agent_skills.py
# Copyright 2025-2026 Arun Rajkumar
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Agent skill management endpoints.

Two parallel access modes:
* **slug-based** — used by the existing SettingsAgentPrompts UI to edit
  one skill at a time; updates fan out across every ``agent_type`` row
  sharing the slug (preserves the original "edit the product-manager
  prompt once" UX after the multi-agent-type split).
* **id-based** — new endpoints for the custom-skill workflow (create,
  delete, set-default).
"""

import uuid

import structlog
from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db, require_permissions
from app.models.agent_skill import AgentSkill, AgentType
from app.models.user import User
from app.repositories.agent_skill import AgentSkillRepository
from app.schemas.agent_skills import (
    AgentSkillRead,
    AgentSkillUpdate,
    CustomSkillCreate,
)
from app.services.skill_loader import (
    Skill,
    list_available_skills,
    load_skill,
    seed_skills_for_org,
)

logger = structlog.get_logger(__name__)

router = APIRouter(tags=["agent-skills"])

_SLUG_PATTERN = r"^[a-z0-9](?:[a-z0-9\-]{0,98}[a-z0-9])?$"
_SLUG_PATH = Path(..., pattern=_SLUG_PATTERN, description="Skill slug (kebab-case)")


def _check_customized(db_prompt: str, slug: str) -> bool:
    """True if the DB prompt differs from the file-based seed default."""
    try:
        file_skill = load_skill(slug)
        return db_prompt != file_skill.prompt
    except (FileNotFoundError, ValueError):
        return True


def _skill_to_read(skill_row: AgentSkill) -> AgentSkillRead:
    """Serialize a DB skill row, including agent_type / default / custom flags."""
    return AgentSkillRead(
        id=skill_row.id,
        skill_slug=skill_row.skill_slug,
        agent_type=skill_row.agent_type,
        is_default=skill_row.is_default,
        is_custom=skill_row.is_custom,
        name=skill_row.name,
        description=skill_row.description,
        tools=skill_row.tools,
        mcp_tools=skill_row.mcp_tools,
        prompt=skill_row.prompt,
        max_turns=getattr(skill_row, "max_turns", 0) or 0,
        timeout_seconds=getattr(skill_row, "timeout_seconds", 0) or 0,
        model=getattr(skill_row, "model", "") or "",
        iteration_model=getattr(skill_row, "iteration_model", "") or "",
        effort=getattr(skill_row, "effort", "") or "",
        is_customized=_check_customized(skill_row.prompt, skill_row.skill_slug),
    )


def _load_skill_or_raise(slug: str) -> Skill:
    """Load a file-defined skill, raising HTTPException on errors."""
    try:
        return load_skill(slug)
    except FileNotFoundError as err:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"Skill not found: {slug}"
        ) from err
    except ValueError as err:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Skill file is malformed for '{slug}': {err}",
        ) from err


@router.get(
    "/",
    response_model=list[AgentSkillRead],
    dependencies=[Depends(require_permissions("settings:view"))],
)
async def list_agent_skills(
    agent_type: AgentType | None = Query(default=None, description="Filter to one agent type"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[AgentSkillRead]:
    """List agent skills for the current org.

    On first call (no rows yet) seeds the file defaults. Optionally
    filters to a single ``agent_type``.
    """
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    skills = await repo.list_all()
    if not skills and list_available_skills():
        try:
            await seed_skills_for_org(current_user.org_id, db)
        except Exception:
            logger.exception("seed_skills_failed", org_id=str(current_user.org_id))
        skills = await repo.list_all()

    if agent_type is not None:
        skills = [s for s in skills if s.agent_type == agent_type]

    return [_skill_to_read(s) for s in skills]


@router.get(
    "/{slug}",
    response_model=AgentSkillRead,
    dependencies=[Depends(require_permissions("settings:view"))],
)
async def get_agent_skill(
    slug: str = _SLUG_PATH,
    agent_type: AgentType | None = Query(
        default=None,
        description="Disambiguate shared slugs (e.g. product-manager)",
    ),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AgentSkillRead:
    """Get one skill by slug; ``agent_type`` is required for shared slugs."""
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    skill_row = await repo.get_by_slug(slug, agent_type=agent_type)
    if skill_row:
        return _skill_to_read(skill_row)

    # Fall back to file default — same shape, no DB row.
    skill = _load_skill_or_raise(slug)
    # File-based skills don't know their agent_type; require the caller
    # to provide it so the response is well-typed.
    if agent_type is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "Skill not in DB yet; pass ?agent_type= to read the file "
                "default with a typed response"
            ),
        )
    return AgentSkillRead.from_skill(slug, skill, agent_type=agent_type)


@router.put(
    "/{slug}",
    response_model=AgentSkillRead,
    dependencies=[Depends(require_permissions("settings:edit"))],
)
async def update_agent_skill(
    body: AgentSkillUpdate,
    slug: str = _SLUG_PATH,
    agent_type: AgentType | None = Query(
        default=None,
        description=(
            "Required when slug is shared across agent types (product-manager, "
            "testing). For unique slugs, the single matching row is updated."
        ),
    ),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AgentSkillRead:
    """Update the (slug, agent_type) row. Partial-update semantics."""
    repo = AgentSkillRepository(db, org_id=current_user.org_id)

    existing = await repo.get_by_slug(slug, agent_type=agent_type)
    if existing is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Skill not found: slug={slug}, agent_type={agent_type}",
        )

    if body.name is not None:
        existing.name = body.name
    if body.description is not None:
        existing.description = body.description
    if body.tools is not None:
        existing.tools = body.tools
    if body.mcp_tools is not None:
        existing.mcp_tools = body.mcp_tools
    if body.prompt is not None:
        existing.prompt = body.prompt
    if body.max_turns is not None:
        existing.max_turns = body.max_turns
    if body.timeout_seconds is not None:
        existing.timeout_seconds = body.timeout_seconds
    if body.model is not None:
        existing.model = body.model
    if body.iteration_model is not None:
        existing.iteration_model = body.iteration_model
    if body.effort is not None:
        existing.effort = body.effort

    await db.flush()
    await db.refresh(existing)

    logger.info(
        "agent_skill_updated",
        slug=slug,
        agent_type=str(existing.agent_type),
        org_id=str(current_user.org_id),
    )
    return _skill_to_read(existing)


@router.delete(
    "/{slug}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_permissions("settings:edit"))],
)
async def reset_agent_skill(
    slug: str = _SLUG_PATH,
    agent_type: AgentType | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Reset a seeded skill back to its file default.

    Custom skills are NOT reset by this endpoint — use ``DELETE /{id}``
    to remove a custom skill instead.
    """
    file_skill = _load_skill_or_raise(slug)
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    existing = await repo.get_by_slug(slug, agent_type=agent_type)
    if existing is None or existing.is_custom:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Seeded skill not found for that slug/agent_type",
        )

    existing.name = file_skill.name
    existing.description = file_skill.description
    existing.tools = file_skill.tools
    existing.mcp_tools = file_skill.mcp_tools
    existing.prompt = file_skill.prompt
    existing.max_turns = file_skill.max_turns
    existing.timeout_seconds = file_skill.timeout_seconds
    existing.model = file_skill.model
    existing.iteration_model = file_skill.iteration_model
    existing.effort = file_skill.effort
    await db.flush()

    logger.info(
        "agent_skill_reset",
        slug=slug,
        agent_type=str(existing.agent_type),
        org_id=str(current_user.org_id),
    )


# ────────────────────────────── Custom skill CRUD ──────────────────────────────


@router.post(
    "/",
    response_model=AgentSkillRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_permissions("settings:edit"))],
)
async def create_custom_skill(
    body: CustomSkillCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AgentSkillRead:
    """Create a user-authored skill tied to ``agent_type``.

    The (slug, agent_type) pair must be unique within the org. Newly
    created skills are NOT marked default — call ``POST /{id}/set-default``
    to promote one explicitly.
    """
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    if await repo.get_by_slug(body.skill_slug, agent_type=body.agent_type):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=(
                f"A skill with slug '{body.skill_slug}' already exists for "
                f"agent_type '{body.agent_type.value}'"
            ),
        )
    created = await repo.create_custom(
        skill_slug=body.skill_slug,
        agent_type=body.agent_type,
        name=body.name,
        description=body.description,
        tools=body.tools,
        mcp_tools=body.mcp_tools,
        prompt=body.prompt,
        max_turns=body.max_turns,
        timeout_seconds=body.timeout_seconds,
        model=body.model,
        iteration_model=body.iteration_model,
        effort=body.effort,
    )
    logger.info(
        "custom_skill_created",
        skill_id=str(created.id),
        slug=created.skill_slug,
        agent_type=str(created.agent_type),
        org_id=str(current_user.org_id),
    )
    return _skill_to_read(created)


@router.post(
    "/{skill_id}/set-default",
    response_model=AgentSkillRead,
    dependencies=[Depends(require_permissions("settings:edit"))],
)
async def set_default_skill(
    skill_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AgentSkillRead:
    """Mark a skill as default for its agent_type (demoting the prior default)."""
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    try:
        updated = await repo.set_default(skill_id)
    except ValueError as err:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(err)) from err
    logger.info(
        "agent_skill_default_set",
        skill_id=str(updated.id),
        agent_type=str(updated.agent_type),
        org_id=str(current_user.org_id),
    )
    return _skill_to_read(updated)


@router.delete(
    "/by-id/{skill_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_permissions("settings:edit"))],
)
async def delete_custom_skill(
    skill_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete a custom skill. Returns 409 for seeded skills."""
    repo = AgentSkillRepository(db, org_id=current_user.org_id)
    existing = await repo.get_by_id(skill_id)
    if existing is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Skill not found")
    if not existing.is_custom:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Seeded skills cannot be deleted; use PUT to reset to file default",
        )
    deleted = await repo.delete_custom(skill_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Skill not found or not custom"
        )
    logger.info(
        "custom_skill_deleted",
        skill_id=str(skill_id),
        org_id=str(current_user.org_id),
    )

File: frontend/src/stores/agentSkills.ts
// Copyright 2025-2026 Arun Rajkumar
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import type { AxiosError } from 'axios'
import api from '@/services/api'

// Mirrors the Python AgentType enum. Frozen string literals keep the
// TS-side type-check honest if a new agent type lands in the backend.
export type AgentType =
  | 'triage'
  | 'bud'
  | 'status'
  | 'standup'
  | 'learning'
  | 'bugLinker'
  | 'reassignment'
  | 'skill'
  | 'techPlan'
  | 'testPlan'
  | 'design'
  | 'slackTriage'

export const AGENT_TYPE_LABELS: Record<AgentType, string> = {
  triage: 'Triage',
  bud: 'BUD (PRD writer)',
  status: 'Status / DevOps',
  standup: 'Standup digest',
  learning: 'Learning recap',
  bugLinker: 'Bug linker',
  reassignment: 'Reassignment',
  skill: 'Skill / code review',
  techPlan: 'Tech plan',
  testPlan: 'Test plan',
  design: 'Design',
  slackTriage: 'Slack triage',
}

export interface AgentSkill {
  id: string | null
  skillSlug: string
  agentType: AgentType
  isDefault: boolean
  isCustom: boolean
  name: string
  description: string
  tools: string[]
  mcpTools: string[]
  prompt: string
  maxTurns: number
  timeoutSeconds: number
  model: string
  iterationModel: string
  effort: string
  isCustomized: boolean
}

interface AgentSkillApi {
  id: string | null
  skill_slug: string
  agent_type: AgentType
  is_default: boolean
  is_custom: boolean
  name: string
  description: string
  tools: string[]
  mcp_tools: string[]
  prompt: string
  max_turns: number
  timeout_seconds: number
  model: string
  iteration_model: string
  effort: string
  is_customized: boolean
}

function fromApi(raw: AgentSkillApi): AgentSkill {
  return {
    id: raw.id ?? null,
    skillSlug: raw.skill_slug,
    agentType: raw.agent_type,
    isDefault: raw.is_default,
    isCustom: raw.is_custom,
    name: raw.name,
    description: raw.description,
    tools: raw.tools,
    mcpTools: raw.mcp_tools,
    prompt: raw.prompt,
    maxTurns: raw.max_turns,
    timeoutSeconds: raw.timeout_seconds ?? 0,
    model: raw.model ?? '',
    iterationModel: raw.iteration_model ?? '',
    effort: raw.effort ?? '',
    isCustomized: raw.is_customized,
  }
}

function extractErrorMessage(err: unknown, fallback: string): string {
  if (err && typeof err === 'object' && 'response' in err) {
    const axiosErr = err as AxiosError<{ detail?: string }>
    if (axiosErr.response?.data?.detail) return axiosErr.response.data.detail
    if (axiosErr.response?.status === 403) return 'You do not have permission for this action.'
  }
  return fallback
}

export interface CustomSkillCreatePayload {
  skillSlug: string
  agentType: AgentType
  name: string
  description?: string
  prompt: string
  tools?: string[]
  mcpTools?: string[]
  maxTurns?: number
  timeoutSeconds?: number
  model?: string
  iterationModel?: string
  effort?: string
}

export const useAgentSkillsStore = defineStore('agentSkills', () => {
  const skills = ref<AgentSkill[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const saving = ref(false)
  const saveSuccess = ref(false)

  // Group view used by the new settings UI — keyed by AgentType, each
  // entry sorted seeded-first then custom alphabetical.
  const byAgentType = computed<Record<string, AgentSkill[]>>(() => {
    const grouped: Record<string, AgentSkill[]> = {}
    for (const s of skills.value) {
      const list = grouped[s.agentType] ?? (grouped[s.agentType] = [])
      list.push(s)
    }
    for (const list of Object.values(grouped)) {
      list.sort((a, b) => {
        if (a.isCustom !== b.isCustom) return a.isCustom ? 1 : -1
        return a.name.localeCompare(b.name)
      })
    }
    return grouped
  })

  async function fetchSkills(): Promise<void> {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<AgentSkillApi[]>('/v1/settings/agent-skills/')
      skills.value = data.map(fromApi)
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to load agent skills.')
    } finally {
      loading.value = false
    }
  }

  async function updateSkill(
    slug: string,
    agentType: AgentType,
    updates: Partial<
      Pick<
        AgentSkill,
        | 'name'
        | 'description'
        | 'tools'
        | 'mcpTools'
        | 'prompt'
        | 'maxTurns'
        | 'timeoutSeconds'
        | 'model'
        | 'iterationModel'
        | 'effort'
      >
    >,
  ): Promise<boolean> {
    saving.value = true
    error.value = null
    saveSuccess.value = false
    try {
      const payload: Record<string, unknown> = {}
      if (updates.name !== undefined) payload.name = updates.name
      if (updates.description !== undefined) payload.description = updates.description
      if (updates.tools !== undefined) payload.tools = updates.tools
      if (updates.mcpTools !== undefined) payload.mcp_tools = updates.mcpTools
      if (updates.prompt !== undefined) payload.prompt = updates.prompt
      if (updates.maxTurns !== undefined) payload.max_turns = updates.maxTurns
      if (updates.timeoutSeconds !== undefined) payload.timeout_seconds = updates.timeoutSeconds
      if (updates.model !== undefined) payload.model = updates.model
      if (updates.iterationModel !== undefined) payload.iteration_model = updates.iterationModel
      if (updates.effort !== undefined) payload.effort = updates.effort

      const { data } = await api.put<AgentSkillApi>(
        `/v1/settings/agent-skills/${slug}`,
        payload,
        { params: { agent_type: agentType } },
      )
      const updated = fromApi(data)
      const idx = skills.value.findIndex(
        s => s.skillSlug === slug && s.agentType === agentType,
      )
      if (idx >= 0) skills.value[idx] = updated
      saveSuccess.value = true
      return true
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to save skill.')
      return false
    } finally {
      saving.value = false
    }
  }

  async function resetSkill(slug: string, agentType: AgentType): Promise<boolean> {
    saving.value = true
    error.value = null
    try {
      await api.delete(`/v1/settings/agent-skills/${slug}`, { params: { agent_type: agentType } })
      await fetchSkills()
      return true
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to reset skill.')
      return false
    } finally {
      saving.value = false
    }
  }

  async function createCustomSkill(payload: CustomSkillCreatePayload): Promise<AgentSkill | null> {
    saving.value = true
    error.value = null
    try {
      const body = {
        skill_slug: payload.skillSlug,
        agent_type: payload.agentType,
        name: payload.name,
        description: payload.description ?? '',
        prompt: payload.prompt,
        tools: payload.tools ?? [],
        mcp_tools: payload.mcpTools ?? [],
        max_turns: payload.maxTurns ?? 0,
        timeout_seconds: payload.timeoutSeconds ?? 0,
        model: payload.model ?? '',
        iteration_model: payload.iterationModel ?? '',
        effort: payload.effort ?? '',
      }
      const { data } = await api.post<AgentSkillApi>('/v1/settings/agent-skills/', body)
      const created = fromApi(data)
      skills.value.push(created)
      saveSuccess.value = true
      return created
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to create custom skill.')
      return null
    } finally {
      saving.value = false
    }
  }

  async function setDefault(skillId: string): Promise<boolean> {
    saving.value = true
    error.value = null
    try {
      const { data } = await api.post<AgentSkillApi>(
        `/v1/settings/agent-skills/${skillId}/set-default`,
      )
      const updated = fromApi(data)
      // Demote any other default in the same agent_type locally, then upsert.
      for (const s of skills.value) {
        if (s.agentType === updated.agentType && s.id !== updated.id) s.isDefault = false
      }
      const idx = skills.value.findIndex(s => s.id === updated.id)
      if (idx >= 0) skills.value[idx] = updated
      saveSuccess.value = true
      return true
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to set default skill.')
      return false
    } finally {
      saving.value = false
    }
  }

  async function deleteCustomSkill(skillId: string): Promise<boolean> {
    saving.value = true
    error.value = null
    try {
      await api.delete(`/v1/settings/agent-skills/by-id/${skillId}`)
      skills.value = skills.value.filter(s => s.id !== skillId)
      return true
    } catch (err) {
      error.value = extractErrorMessage(err, 'Failed to delete custom skill.')
      return false
    } finally {
      saving.value = false
    }
  }

  return {
    skills,
    byAgentType,
    loading,
    error,
    saving,
    saveSuccess,
    fetchSkills,
    updateSkill,
    resetSkill,
    createCustomSkill,
    setDefault,
    deleteCustomSkill,
  }
})
